hazelcast/src/main/java/com/hazelcast/cp/internal/RaftService.java

```diff
@@ -1336,6 +1380,59 @@ public void applyUnsafeModeState(int partitionId, UnsafeModePartitionState state
         unsafeModeStates[partitionId].apply(state);
     }
 
+    public UUID registerMembershipListener(CPMembershipListener listener) {
+        return nodeEngine.getEventService().registerListener(SERVICE_NAME, EVENT_TOPIC_MEMBERSHIP, listener).getId();
+    }
+
+    public boolean removeMembershipListener(UUID id) {
+        return nodeEngine.getEventService().deregisterListener(SERVICE_NAME, EVENT_TOPIC_MEMBERSHIP, id);
+    }
+
+    public UUID registerAvailabilityListener(CPGroupAvailabilityListener listener) {
+        return nodeEngine.getEventService().registerListener(SERVICE_NAME, EVENT_TOPIC_AVAILABILITY, listener).getId();
+    }
+
+    public boolean removeAvailabilityListener(UUID id) {
+        return nodeEngine.getEventService().deregisterListener(SERVICE_NAME, EVENT_TOPIC_AVAILABILITY, id);
+    }
+
+    @Override
+    public void dispatchEvent(Object e, EventListener l) {
+        long now = Clock.currentTimeMillis();
+        recentAvailabilityEvents.values().removeIf(expirationTime -> expirationTime < now);
+
+        if (e instanceof CPMembershipEvent) {
+            CPMembershipEvent event = (CPMembershipEvent) e;
+            CPMembershipListener listener = (CPMembershipListener) l;
+            switch (event.getType()) {
+                case ADDED:
+                    listener.memberAdded(event);
+                    break;
+                case REMOVED:
+                    listener.memberRemoved(event);
+                    break;
+                default:
+                    throw new IllegalArgumentException("Unhandled event: " + event);
```

Comment:
If we throw an exception here, the availability event might be missing.

Reply:
At this point we already know event object is instance of `CPMembershipEvent`. This `default` case should be unreachable under normal conditions, it's just for the case we add a new `CPMembershipEvent` type.